Draw area exit names

gang929 opened the pull request:
When a room connect to another area's room, draw the exit area name on the line end

<!-- Keep the title short & concise so your mom can understand it -->
#### Brief overview of PR changes/additions

#### Motivation for adding to Mudlet

#### Other info (issues closed, discussion etc)
To not show area names in a room:

```
-- don't show exit:
setRoomUserData(roomid, "room.ui_dontShowAreaExitName", "yes")

-- move normal exit a little bit
setRoomUserData(roomid, "room.ui_areaExitNameOffset.south", "20")
setRoomUserData(roomid, "room.ui_areaExitNameOffset.south", "20 20")
-- move special exit a little bit
setRoomUserData(roomid, "room.ui_areaExitNameOffset.northdown", "20")
setRoomUserData(roomid, "room.ui_areaExitNameOffset.northup", "20 20")
```

gang929 committed:
draw AreaExitName

When a room connect to another area's room, draw the exit area name on the line end

SlySven reviewed (commented):
😄 I like how the area name for a custom exit line for a special exit gets positioned in front of the arrow.

☹️ You need to compensate for the need to scale the offset position for the text at different zooms - note how the normal area exit is too close to the room when zoomed in for the normal exit direction without a custom exit line. You probably want to examine how the position for the tip of the arrow for the area exit is calculated and deduce which variable factors for that at different zoom amount...
![image](https://user-images.githubusercontent.com/6163092/99013294-0fed1200-2548-11eb-846e-3d75fa53fb1b.png)
![image](https://user-images.githubusercontent.com/6163092/99013323-1ed3c480-2548-11eb-8c13-52f69eeb556c.png)

☹️ I am puzzled why a custom exit line for a *normal* direction exit does NOT get the destination area name?

☹️ This feature is turned on for every area exit (well, except for the case of the previous paragraph) - however it is possible that previous map makers may well have done something with map labels and custom exit lines themselves already - so that this will be redundant/duplicate things for some users. I think we may need to consider whether we need control over it and whether that should be per area exit or more globally...

It would be helpful if the new code could match up with the code style of our existing code - the main issue here would be the placement (or rather absence of) `{`...`}` around the conditionally executed code in `if (...)` statements  - we really like an opening `{` at the end of the `if` condition line and a closing `}` on its own line (unless followed by an ` else {`... even for single lines of conditional code. 🙏

gang929 committed:
fix code style

gang929 committed:
fix exit name postion with zoom

gang929 commented:
> why a custom exit line for a normal direction exit does NOT get the destination area name?

Should normal direction exit use normal exit line, not a special (custom) exit line ?

> This feature is turned on for every area exit 

Is add a map userdata or a area userdata to ONOFF the feature?

vadi2 commented:
> Is add a map userdata or a area userdata to ONOFF the feature?

I think room userdata is good, similar to how we did it for room names (https://wiki.mudlet.org/w/Mapping_script#Labelling_rooms). I think this is a good option to have on by default, so how about `room.ui_dontShowAreaExitName`?

@Mudlet/lua-interface @demonnic @Kebap @keneanung 

> fix exit name postion with zoom

Zoom is good, but at some point it doesn't work anymore, maybe cap it?

![zoom-too-much](https://user-images.githubusercontent.com/110988/99140999-c800e480-2646-11eb-86da-4b467aa917e8.gif)

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -1993,6 +1993,58 @@ void T2DMap::paintAreaExits(QPainter& painter, QPen& pen, QList<int>& exitList,
+                        }
+                    }
+                    if (rID > -1) {
+                        TRoom* pE = mpMap->mpRoomDB->getRoom(rID);
```
```suggestion
                        auto exitRoom = mpMap->mpRoomDB->getRoom(rID);
```

SlySven replied:
I'd not like to loose track that it is a pointer and not a TRoom instance - my preference would have been:
```diff
-                        TRoom* pE = mpMap->mpRoomDB->getRoom(rID);
+                        auto pExitRoom = mpMap->mpRoomDB->getRoom(rID);
```
🤷‍♂️

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -1993,6 +1993,58 @@ void T2DMap::paintAreaExits(QPainter& painter, QPen& pen, QList<int>& exitList,
                     }
                     painter.drawPolyline(polyLinePoints.data(), polyLinePoints.size());
 
+                    int rID = -1;
```
```suggestion
                    int exitRoomId = -1;
```

ryanstadther commented:
Would there be any ability to offset the text using an X,Y offset? Similar to how custom exit lines are set up?  I think there are a lot of places where the exit name would overlap with other rooms but it could look nice if the name was offset a bit from the default position. Just a nice to have if it isn't too difficult.

vadi2 commented:
Good idea, could be same style as https://wiki.mudlet.org/w/Mapping_script#Labelling_rooms via room user data

SlySven commented:
> >  why a custom exit line for a normal direction exit does NOT get the destination area name?

> Should normal direction exit use normal exit line, not a special (custom) exit line ?

A custom exit line can be applied to ANY exit (normal or special) - it is just that Special Exits can ONLY be shown on the (2D) map with a custom exit line as otherwise there is NO indication that they are there.

gang929 committed:
rename rID -> exitRoomId, pE -> pExitRoom

gang929 committed:
Add a option room.ui_dontShowAreaExitName for ROOM user data with default ON to ONOFF show area exit name feature. The default behaviour is show, call setRoomUserData(roomid, "room.ui_dontShowAreaExitName", "true") to hide it.

gang929 commented:
> Zoom is good, but at some point it doesn't work anymore, maybe cap it?

How do it?  When zoom is less some value, hide it?

gang929 commented:
> Would there be any ability to offset the text using an X,Y offset? Similar to how custom exit lines are set up? I think there are a lot of places where the exit name would overlap with other rooms but it could look nice if the name was offset a bit from the default position. Just a nice to have if it isn't too difficult.

The X,Y offset may be should have a difference about zoom value, but it's not has a perfect ALG to set it

gang929 reviewed (commented):
The build passed again...

vadi2 commented:
Yeah, I rebuilt it manually. I don't think Travis allows you to restart it, unfortunately.

vadi2 commented:
> How do it? When zoom is less some value, hide it?

Yeah. When it doesn't look good / doesn't make sense anymore.

> The X,Y offset may be should have a difference about zoom value, but it's not has a perfect ALG to set it

Yes, we had same problem with room names - solution was picking a good default + allowing to set custom offset via userdata, see `room.ui_nameOffset`

gang929 committed:
When the scaling ratio is less than 1 times and the exit area name rect width is greater than 1 / 3 of the area width, don't show the exit area name.

gang929 committed:
add a option room.ui_areaExitNameOffset for ROOM user data, set it by EXITDIR, when set the offset, area exit name will display with the offset value. The offset value is "Yoffset" or "Xoffset Yoffset", for example:
```
-- normal exit
setRoomUserData(roomid, "room.ui_areaExitNameOffset.south", "20")
setRoomUserData(roomid, "room.ui_areaExitNameOffset.south", "20 20")
-- special exit
setRoomUserData(roomid, "room.ui_areaExitNameOffset.northdown", "20")
setRoomUserData(roomid, "room.ui_areaExitNameOffset.northup", "20 20")
```

SlySven commented on src/T2DMap.cpp:
```diff
@@ -2001,6 +2029,7 @@ void T2DMap::paintAreaExits(QPainter& painter, QPen& pen, QList<int>& exitList,
                     painter.drawPolyline(polyLinePoints.data(), polyLinePoints.size());
 
                     if (!getUserDataBool(room->userData, ROOM_UI_DONTSHOWAREAEXITNAME, false)) {
+                        QPointF areaExitNameOffset = parseOffsetString(room->userData.value(ROOM_UI_AREAEXITNAMEOFFSET + "." + itk.key()));
```
These things should be done with a `QStringLiteral(`...`)` wrapper so that you are not working on raw string literals - and are explicitly marking them as internal non UI texts that are NOT to be put through the translation system that UI text should be...

In this case you should (*I* think) use:
```cpp
    QPointF areaExitNameOffset = parseOffsetString(room->userData.value(QStringLiteral(ROOM_UI_AREAEXITNAMEOFFSET ".%1").arg(itk.key())));
```

SlySven commented:
🤔 I must try this one again to see what happens if the area exit is in a non-XY plane but still a "*normal*" direction - i.e. *up*, *down*, *in* or *out*...

vadi2 commented on src/T2DMap.cpp:
```diff
@@ -1993,6 +2028,63 @@ void T2DMap::paintAreaExits(QPainter& painter, QPen& pen, QList<int>& exitList,
                     painter.drawPolyline(polyLinePoints.data(), polyLinePoints.size());
 
+                    if (!getUserDataBool(room->userData, ROOM_UI_DONTSHOWAREAEXITNAME, false)) {
+                        QPointF areaExitNameOffset = parseOffsetString(room->userData.value(ROOM_UI_AREAEXITNAMEOFFSET + "." + itk.key()));
```
```suggestion
                        QPointF areaExitNameOffset = parseOffsetString(room->userData.value(QStringLiteral(ROOM_UI_AREAEXITNAMEOFFSET ".%1").arg(itk.key())));
```

vadi2 reviewed (approved):
@gang929 looks good in general. Sorry for the delay, in merging this, I hope this doesn't put you off future contributions.

Are you comfortable with documenting it on https://wiki.mudlet.org/w/Mapping_script?

gang929 commented:
@vadi2, How to merge to master or development branch ?

vadi2 commented:
Try this:

![image](https://user-images.githubusercontent.com/110988/125641886-2a138509-7930-430f-b681-ebb160fc0a2a.png)

vadi2 commented:
This still needs to be documented, we missed doing this 😮
